libraries/standard/mqtt/src/mqtt.c

```diff
@@ -199,6 +203,79 @@ static MQTTStatus_t receiveConnack( MQTTContext_t * const pContext,
 
+/**
+ * @brief Resends pending acks for a re-established MQTT session.
+ *
+ * @param[in] pContext Initialized MQTT context.
+ *
+ * @return #MQTTSendFailed if transport send failed;
+ * #MQTTSuccess otherwise.
+ */
+static MQTTStatus_t resendPendingAcks( MQTTContext_t * const pContext );
+
+/**
+ * @brief Resends PUBREL for the outgoing QoS2 publish and updates
+ * the state machine based on the state of the packet.
+ *
+ * @param[in] pContext Initialized MQTT context.
+ * @param[in] state Current state of the packet.
+ *
+ * @return #MQTTSendFailed if transport send failed;
+ * #MQTTSuccess otherwise.
+ */
+static MQTTStatus_t resendPubrelForPacketInState( MQTTContext_t * const pContext,
+                                                  MQTTPublishState_t state );
+
+/**
+ * @brief Checks if an incoming publish was received before.
+ *
+ * @param[in] pContext Initialized MQTT context.
+ * @param[in] packetId Packet identifier of the incoming publish packet.
+ * @param[in] qos Qos level of the incoming publish packet.
+ * @param[out] pIsStateUpdateNeededAfterAck Whether a state update is required
+ * after sending an ack for the incoming publish.
+ * @param[out] pPublishRecordState Publish record state for the ack to be sent.
+ *
+ * @return true if publish was already received and present in the state records;
+ * false otherwise.
+ */
+static bool checkIfPublishWasReceivedBefore( MQTTContext_t * const pContext,
+                                             uint16_t packetId,
+                                             MQTTQoS_t qos,
+                                             bool * pIsStateUpdateNeededAfterAck,
+                                             MQTTPublishState_t * pPublishRecordState );
+
+/**
+ * @brief Checks if packet with a given state is present in state records.
+ *
+ * @param[in] pContext Initialized MQTT context.
+ * @param[in] packetId Packet identifier to query in the state records.
+ * @param[in] state Packet state to query for in the state records.
+ *
+ * @return true if packet id with a given state is present in the state records;
+ * false otherwise.
+ */
+static bool isPacketStatePresentInStateRecords( MQTTContext_t * const pContext,
+                                                uint16_t packetId,
+                                                MQTTPublishState_t state );
+
+/**
+ * @brief Checks if a PUBREL packet was received before.
+ *
+ * @param[in] pContext Initialized MQTT context.
+ * @param[in] packetId Packet identifier of the incoming publish packet.
+ * @param[out] pIsStateUpdateNeededAfterAck Whether a state update is required
+ * after sending an ack for the incoming PUBREL.
+ * @param[out] pPublishRecordState Publish record state for the ack to be sent.
+ *
+ * @return true if PUBREL was already received and present in the state records;
+ * false otherwise.
+ */
+static bool checkIfPubrelWasReceivedBefore( MQTTContext_t * const pContext,
```

Comment:
This looks like a very specific function that could very easily be generalized to handle PUBREC as well. Do we not care about it?

Reply:
Only PUBLISH and PUBREL packets will be resend as per the spec. http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718103